### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by marcin-golebiowski**

### CODE REVIEW

**Comment by marcin-golebiowski** [comment]
@svenboulanger Could you please check why after my changes tests are randomly failing ?

**Comment by svenboulanger** [comment]
Check the development pull request for that and see if it is the same test that fails. It has to do with the order of entities (by their hash code).

**Comment by marcin-golebiowski** [comment]
@svenboulanger SpiceSharpParser tests fail randomly  when I'm using development branch.
I agree with you that it's because of the EntityPriorityComparer and using GetHashCode. I've modified it in the following way:

image

However after that I have problem with Priority. Please change its setter to public. So I can write some hacks about it.

**Comment by svenboulanger** [comment]
Sorting by name won't get rid of the underlying problem. The entity order should not influence the simulation results, this may be a bug somewhere else.

File: SpiceSharp/General/Parameterized.cs [hunk 1]
```diff
@@ -20,8 +26,55 @@ protected IEnumerable<MemberInfo> Members
...
+            }
+        }
+
+        /// <summary>
+        /// Gets all members in the class.
+        /// </summary>
+        /// <value>
+        /// The members.
+        /// </value>
+        protected IEnumerable<Tuple<MemberInfo, List<Attribute>>> MembersExt
```
[the remark this answers is not in the document]
Comment (marcin-golebiowski) [reply SpiceSharp/General/Parameterized.cs]: @svenboulanger  Sure. I will  fix it.

File: SpiceSharp/Circuits/Entity.cs [hunk 2]
```diff
@@ -24,7 +24,10 @@ public abstract class Entity
         protected static void RegisterBehaviorFactory(Type entityType, BehaviorFactoryDictionary dictionary)
         {
             // We do this to avoid anyone unregistering factories!
-            BehaviorFactories.Add(entityType, dictionary);
+            lock (BehaviorFactories)
+            {
+                BehaviorFactories.Add(entityType, dictionary);
```
[the remark this answers is not in the document]
Comment (marcin-golebiowski) [reply SpiceSharp/Circuits/Entity.cs]: @svenboulanger  It's old code. I've changed that later.

File: SpiceSharp/General/Parameterized.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (svenboulanger) [reply SpiceSharp/General/Parameterized.cs]: I meant to change the property name to eg. `CachedMembers`.

[the remark this answers is not in the document]
Comment (marcin-golebiowski) [reply SpiceSharp/General/Parameterized.cs]: @svenboulanger  I will fix that.

[the remark this answers is not in the document]
Comment (marcin-golebiowski) [reply SpiceSharp/General/Parameterized.cs]: @svenboulanger  Fixed

**Pull request merged**